Xamarin.PropertyEditing/ViewModels/ResourceSelectorViewModel.cs

```diff
@@ -0,0 +1,144 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Xamarin.PropertyEditing.ViewModels
+{
+	internal class ResourceSelectorViewModel
+		: NotifyingObject
+	{
+		public ResourceSelectorViewModel (IResourceProvider provider, IEnumerable<object> targets, IPropertyInfo property)
+		{
+			if (provider == null)
+				throw new ArgumentNullException (nameof (provider));
+			if (targets == null)
+				throw new ArgumentNullException (nameof (targets));
+			if (property == null)
+				throw new ArgumentNullException (nameof (property));
+
+			Provider = provider;
+			this.targets = targets.ToArray();
+			Property = property;
+			UpdateResources();
+
+			this.resourcesView = new SimpleCollectionView (this.resources, new SimpleCollectionViewOptions {
+				DisplaySelector = (o) => ((Resource)o).Name
+			});
+		}
+
+		public IPropertyInfo Property
+		{
+			get;
+		}
+
+		public IResourceProvider Provider
+		{
+			get;
+		}
+
+		public IEnumerable Resources => this.resourcesView;
+
+		public bool IsLoading
+		{
+			get { return this.isLoading; }
+			set
+			{
+				if (this.isLoading == value)
+					return;
+
+				this.isLoading = value;
+				OnPropertyChanged();
+			}
+		}
+
+		public bool ShowSystemResources
+		{
+			get {  return this.showSystemResources; }
+			set
+			{
+				if (this.showSystemResources == value)
+					return;
+
+				this.showSystemResources = value;
+				OnPropertyChanged();
+				UpdateResourceFilter();
+			}
+		}
+
+		public string FilterText
+		{
+			get { return this.filterText; }
+			set
+			{
+				if (this.filterText == value)
+					return;
+
+				this.filterText = value;
+				OnPropertyChanged();
+				UpdateResourceFilter();
+			}
+		}
+
+		private readonly ObservableCollectionEx<Resource> resources = new ObservableCollectionEx<Resource>();
+		private readonly SimpleCollectionView resourcesView;
+		private bool showSystemResources = true, isLoading;
+		private string filterText;
+		private readonly object[] targets;
+
+		private void UpdateResourceFilter()
+		{
+			if (String.IsNullOrWhiteSpace (FilterText) && ShowSystemResources) {
+				this.resourcesView.Options.Filter = null;
+				return;
+			}
+
+			this.resourcesView.Options.Filter = ResourceFilter;
+		}
+
+		private bool ResourceFilter (object item)
```

Comment:
Why isn't that taking `Resource` since you're going to immediately cast it anyway?

Reply:
Because that would just move the cast somewhere else and create a second method in the process.

Reply:
It would move it arguably where it belongs, with a safer result: we're trading compile-time safety for runtime exceptions. I'm not following why this would create a second method.

Reply:
> we're trading compile-time safety for runtime exceptions.

No we're not, we'll have to cast it regardless, `SimpleCollectionView` is not generic remember?

> I'm not following why this would create a second method.

3 lines up.

Reply:
Well, it's private anyways.